File: scripts/collect-americas-supplemental-sources.mjs
#!/usr/bin/env node
import fs from 'node:fs/promises';
import crypto from 'node:crypto';
import path from 'node:path';
import process from 'node:process';

const arg=name=>{const i=process.argv.indexOf(name);return i>=0?process.argv[i+1]:null;};
const project=path.resolve(arg('--project')||'');
const manifestPath=path.resolve(arg('--manifest')||'data/registries/americas-supplemental-sources.json');
if(!project)throw new Error('--project is required');
const manifest=JSON.parse(await fs.readFile(manifestPath,'utf8'));
const root=path.join(project,'raw','supplemental-country-sources');await fs.mkdir(root,{recursive:true});
const maxBytes=100*1024*1024,timeoutMs=30000,receipts=[];
const extension=(type,url)=>{const pathname=new URL(url).pathname.toLowerCase();for(const ext of ['.xlsx','.csv','.geojson','.json','.pdf','.zip','.html'])if(pathname.endsWith(ext))return ext;if(type.includes('pdf'))return '.pdf';if(type.includes('json'))return '.json';if(type.includes('csv'))return '.csv';return '.html';};
for(const source of manifest.sources||[]){
  const receipt={...source,requested_at:new Date().toISOString(),status:'failed_with_evidence'};
  try{
    const response=await fetch(source.url,{redirect:'follow',signal:AbortSignal.timeout(timeoutMs),headers:{'user-agent':'AreaData source audit/0.10 (+https://areadata.net/)','accept':'text/html,application/pdf,application/json,text/csv,*/*'}});
    receipt.http_status=response.status;receipt.final_url=response.url;receipt.content_type=response.headers.get('content-type')||'';
    if(!response.ok)throw new Error(`HTTP ${response.status}`);
    const chunks=[];let size=0;for await(const chunk of response.body){size+=chunk.length;if(size>maxBytes)throw new Error(`Response exceeds ${maxBytes} bytes`);chunks.push(chunk);}const body=Buffer.concat(chunks);
    const dir=path.join(root,source.country_area_id);await fs.mkdir(dir,{recursive:true});const file=path.join(dir,`${source.domain}-${source.id}${extension(receipt.content_type,response.url)}`);
    await fs.writeFile(file,body);receipt.status='acquired';receipt.bytes=body.length;receipt.sha256=crypto.createHash('sha256').update(body).digest('hex');receipt.path=path.relative(project,file).replaceAll('\\','/');
  }catch(error){receipt.error=`${error.name}: ${error.message}`;}
  receipt.completed_at=new Date().toISOString();receipts.push(receipt);console.log(`${source.country_area_id} ${source.id} ${receipt.status}`);
}
const output={schema_version:'1.0',generated_at:new Date().toISOString(),manifest:path.relative(process.cwd(),manifestPath).replaceAll('\\','/'),receipts};
await fs.writeFile(path.join(root,'receipt.json'),JSON.stringify(output,null,2)+'\n');
console.log(JSON.stringify({acquired:receipts.filter(row=>row.status==='acquired').length,failed:receipts.filter(row=>row.status!=='acquired').length},null,2));

File: scripts/collect-americas-supplemental-sources.py
#!/usr/bin/env python3
"""Fetch a reviewed supplemental-source manifest with bounded, hashed receipts."""

from __future__ import annotations

import argparse
import hashlib
import json
import mimetypes
import re
import urllib.error
import urllib.request
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlparse


def extension(content_type: str, url: str) -> str:
    suffix = Path(urlparse(url).path).suffix.lower()
    if suffix in {".xlsx", ".xls", ".csv", ".geojson", ".json", ".pdf", ".zip", ".html"}:
        return suffix
    guessed = mimetypes.guess_extension(content_type.split(";", 1)[0].strip())
    return ".html" if guessed in {None, ".htm"} else guessed


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--project", required=True, type=Path)
    parser.add_argument("--manifest", default="data/registries/americas-supplemental-sources.json", type=Path)
    parser.add_argument("--timeout", default=45, type=int)
    parser.add_argument("--max-bytes", default=100 * 1024 * 1024, type=int)
    args = parser.parse_args()
    project = args.project.resolve()
    manifest = json.loads(args.manifest.read_text(encoding="utf-8"))
    root = project / "raw/supplemental-country-sources"
    root.mkdir(parents=True, exist_ok=True)
    receipts = []
    for source in manifest.get("sources", []):
        receipt = {**source, "requested_at": datetime.now(timezone.utc).isoformat(), "status": "failed_with_evidence"}
        try:
            request = urllib.request.Request(source["url"], headers={"User-Agent": "Mozilla/5.0 AreaData source audit/0.10", "Accept": "text/html,application/pdf,application/json,text/csv,*/*"})
            with urllib.request.urlopen(request, timeout=args.timeout) as response:
                body = response.read(args.max_bytes + 1)
                if len(body) > args.max_bytes:
                    raise ValueError(f"Response exceeds {args.max_bytes} bytes")
                receipt.update({"http_status": response.status, "final_url": response.url, "content_type": response.headers.get("Content-Type", "")})
            country_dir = root / source["country_area_id"]
            country_dir.mkdir(parents=True, exist_ok=True)
            safe_id = re.sub(r"[^A-Za-z0-9._-]+", "_", source["id"])
            target = country_dir / f"{source['domain']}-{safe_id}{extension(receipt['content_type'], receipt['final_url'])}"
            target.write_bytes(body)
            receipt.update({"status": "acquired", "bytes": len(body), "sha256": hashlib.sha256(body).hexdigest(), "path": target.relative_to(project).as_posix()})
        except Exception as exc:
            receipt["error"] = f"{type(exc).__name__}: {exc}"
        receipt["completed_at"] = datetime.now(timezone.utc).isoformat()
        receipts.append(receipt)
        print(source["country_area_id"], source["id"], receipt["status"])
    output = {"schema_version": "1.0", "generated_at": datetime.now(timezone.utc).isoformat(), "manifest": args.manifest.as_posix(), "receipts": receipts}
    (root / "receipt.json").write_text(json.dumps(output, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps({"acquired": sum(row["status"] == "acquired" for row in receipts), "failed": sum(row["status"] != "acquired" for row in receipts)}, indent=2))


if __name__ == "__main__":
    main()
